### ISSUE TITLE
unique_everseen should be templated by a hash function and use a different set object

### ISSUE THREAD

**Issue by nimrodfg** [ryanhaining/cppitertools#90]
Currently we can't use unique_everseen for proto-buffs, or any element that doesn't have std::hash. 
Would make sense to make functions for which its relevant to be templated by equality and hash operators.

**Comment by ryanhaining** [comment]
Good point. I added an overload. I'm not sure if there's a good/right way to get the pipe syntax working with the new version though.
